### PR TITLE
#1993 integrating thunknyc profile to cider.

### PR DESCRIPTION
**Pull request by jmayaalv**
This is my attempt to solve https://github.com/clojure-emacs/cider/issues/1993#issuecomment-354975815

I am still having problems displaying the information on cider. I will need some guidance. 

Before submitting a PR make sure the following things have been done:

- [ ] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [x] You've added tests to cover your change(s)
- [x ] All tests are passing
- [x] The new code is not generating reflection warnings
- [x ] You've updated the readme (if adding/changing middleware)

Thanks!

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -9,6 +9,7 @@
   :dependencies [[org.clojure/tools.nrepl "0.2.13"]
                  [org.tcrawley/dynapath "1.0.0"]
                  [cider/orchard "0.1.0-SNAPSHOT"]
+                 [thunknyc/profile "0.5.2"]
```
Comment (bbatsov) [inline project.clj]: This should be a source dep.

File: README.md [hunk 2]
```diff
@@ -157,6 +157,7 @@ Middleware        | Op(s)      | Description
 `wrap-spec`         | `spec-list/spec-form/spec-example` | Spec browsing.
 `wrap-pprint`     | | Adds pretty-printing support to code evaluation. It also installs a dummy `pprint-middleware` op. Thus `wrap-pprint` is discoverable through the `describe` op.
 `wrap-pprint-fn`  | | Provides a common pretty-printing interface for other middlewares that need to perform customisable pretty-printing.
+`wrap-profile`    | Provides Profiling support for [Profiling](https://github.com/thunknyc/profile)
```
Comment (bbatsov) [inline README.md]: You should list here the ops provided by the middleware.

**Comment by bbatsov** [comment]
You need to rebase this on top of `master`. I would also be nice if you added a few tests similar to the ones for the tracing middleware.

File: src/cider/nrepl/middleware/profile.clj [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+       (catch Exception e  (send-exception e msg transport))))
+
+(defn handle-profile
+  [handler msg]
+  (let [{:keys [op] } msg]
+    (case op
+      "toggle-profile"
+      (toggle-profile msg)
+      "toggle-profile-ns"
+      (toggle-profile-ns msg)
```
Comment (ckoparkar) [inline src/cider/nrepl/middleware/profile.clj]: Great work @jmayaalv!

One minor comment; should the formatting of the case expression be this instead: 
```
(case op
  "toggle-profile" (toggle-profile msg)
  "..."  ...)
```

Would be a bit nicer to read I think.

**Comment by jmayaalv** [comment]
Thank you! Tests Added, Readme updated and cleaned up the formatting  as per @ckoparkar suggestion.

**Comment by bbatsov** [comment]
Hmm, one test seems to be failing for just one build, and it seems unrelated to the PR. I re-ran this a few times and it always fails. @gonewest818 @xiongtx Can you take a look a this?

**Comment by xiongtx** [comment]
Re-ran the build and [that particular combination just passed](https://travis-ci.org/clojure-emacs/cider-nrepl/jobs/338104228). Chalk it up to @travis-ci problems, I guess.

**Comment by bbatsov** [comment]
👍

**Comment by xiongtx** [comment]
I noticed that this code (as well as the Elisp code for `cider-profile.el` in clojure-emacs/cider#2197) is taken almost verbatim from https://github.com/thunknyc/profile.

Aside from the [EPL](https://github.com/thunknyc/nrepl-profile/blob/9aa7e404f53f5136ac8d15301ce8ca2924785cb3/LICENSE) being [incompatible with the GPL](https://www.eclipse.org/legal/eplfaq.php#GPLCOMPATIBLE), I see no attribution to @edw, which is not cool in my :book:.

**Comment by gonewest818** [comment]
If @edw is available, the fastest resolution might be to ask him to dual-license the original code (as GPL **and** EPL).  That would be possible only if he retains all the copyright, and assuming he himself hasn't adopted from a third party that can't or won't consent to the licensing change.

**Comment by xiongtx** [comment]
Hmm, the `cider-profile.el` in `nrepl-profile` actually contains the GPL [in its source code](https://github.com/thunknyc/nrepl-profile/blob/9aa7e404f53f5136ac8d15301ce8ca2924785cb3/cider-profile.el#L10-L21). So we can include it in CIDER w/out problem. _However_, we definitely need to [attribute @edw](https://github.com/thunknyc/nrepl-profile/blob/9aa7e404f53f5136ac8d15301ce8ca2924785cb3/cider-profile.el#L2).

Which leads us to the broader issue that this code was >95% taken from someone else, to whom no attribution is given.

**Comment by bbatsov** [comment]
> Which leads us to the broader issue that this code was >95% taken from someone else, to whom no mention is given.

I just didn't pay attention to this. We should certainly mention him as the author. To my knowledge he developed this library mostly to be able to use it from CIDER, so I doubt he'll mind promoting his code upstream.

**Comment by bbatsov** [comment]
@xiongtx I've added the missing attributions, we should good now.

**Comment by edw** [comment]
Hey folks, yeah, I’d be happy to see profile integrated. @bbatsov is absolutely correct: I wrote it with CIDER integration in mind. I’ll add a note to the repo that it’s been absorbed into CIDER.

**Pull request merged**